backend/hidrive/helpers.go

```diff
@@ -0,0 +1,902 @@
+// starting at the given offset.
+// Returns the amount of successfully written bytes.
+// If nothing was written this will be 0.
+//
+// Replaces the file contents starting from the given byte offset
+// with the content of the io.Reader.
+// If the offset is beyond the file end, the file is extended up to the offset.
+//
+// The upload is done multiple chunks of the given size in parallel.
+// Therefore this can and be used to upload files of any size efficiently.
+// The number of parallel transfers is limited by transferLimit which should larger than 0.
+//
+// NOTE: Because it is inefficient to set the modification time with every chunk,
+// setting it to a specific value must be done in a separate request
+// after this operation finishes.
+//
+// NOTE: This method uses patchFile and may create sparse files,
+// especially if the upload of a chunk fails unexpectedly.
+// See note about sparse files in patchFile.
+// If any of the uploads fail, the process will be aborted and
+// the first error that occurred will be returned.
+// This is not an atomic operation,
+// therefore if the upload fails the file may be partially modified.
+//
+// This returns fs.ErrorObjectNotFound if the object is not found.
+func (f *Fs) updateFileChunked(ctx context.Context, path string, content io.Reader, offset uint64, chunkSize int, transferLimit int64) (amount uint64, err error) {
+	amount = uint64(0)
+	mutex := new(sync.Mutex)
+	waitgroup, wgCtx := errgroup.WithContext(ctx)
+	transferSemaphore := semaphore.NewWeighted(transferLimit)
+
+	var readErr error
+	startMoreTransfers := true
+	zeroTime := time.Time{}
+	for chunk := uint64(0); startMoreTransfers; chunk++ {
+		// Acquire semaphore to limit number of transfers in parallel.
+		readErr = transferSemaphore.Acquire(wgCtx, 1)
+		if readErr != nil {
+			break
+		}
+
+		// Read a chunk of data.
+		chunkReader, bytesRead, readErr := readerForChunk(content, chunkSize)
+		if bytesRead < chunkSize {
+			startMoreTransfers = false
+		}
+		if readErr != nil || bytesRead <= 0 {
+			break
+		}
+
+		// Transfer the chunk.
+		contentOffset := uint64(chunkSize) * chunk
+		chunkOffset := contentOffset + offset
+		waitgroup.Go(func() error {
+			// After this upload is done,
+			// signal that another transfer can be started.
+			defer transferSemaphore.Release(1)
+			uploadErr := f.patchFile(wgCtx, path, cachedReader(chunkReader), chunkOffset, zeroTime)
+			// Safely update lower-bound
+			// for amount of bytes written before failures.
+			// The errgroup-context wgCtx is canceled implicitly
+			// as soon as the first routine returns an error.
+			if uploadErr == nil {
+				mutex.Lock()
+				select {
+				case <-wgCtx.Done():
+					// Another upload failed already.
+					// Nothing more to do.
+				default:
+					contentPosition := contentOffset + uint64(bytesRead)
+					fs.Debugf(f, "Done uploading chunk of size %v at offset %v.", bytesRead, chunkOffset)
+					if amount < contentPosition {
```

Comment:
I'm not sure this calculation on `amount` is correct if it is intended to be the last written block position.

What if block 1 was OK, block 2 failed and block 3 was OK. This could easily complete in the order block 1 OK, block 3 OK, block 2 FAIL. Won't amount be set wrong in this case? As in set to end of block 3 where there is a gap for block 2?

Or is it just the high water mark for correctly written blocks?

Or is it the high water mark for any blocks written, correctly or not.

Either way I think you can tidy this if statement up by dealing with `amount` first if uploadErr == nil.

You might want to check out the s3 multipart uploader for ideas - this code has been through many iterations to get the concurrency right!

https://github.com/rclone/rclone/blob/5c6a958ad88db85471fc8d1975a5378f67e16fc1/backend/s3/s3.go#L4085

Reply:
> block 1 OK, block 3 OK, block 2 FAIL

Hm, I don't see anything wrong in this scenario. The desired result would be that `amount` points to the end of block 1.
I think this is exactly what the code is doing:
0. `amount = 0`
1. block 1 OK -> no upload error & no other upload failed -> `amount < end_block1` -> `amount = end_block1`
2. block 3 OK -> no upload error & no other upload failed -> `amount < end_block3` -> `amount = end_block3`
3. block 2 FAIL -> upload error & no other upload failed -> `amount = start_block2 (== end_block1)`

Reply:
> Either way I think you can tidy this if statement up by dealing with amount first if uploadErr == nil.

Not sure what you mean by this, could you elaborate? :)